mcode/rems-smart-on-fhir
# Migrate from react-app-rewired and jest to vite and vitest

## Describe your changes

> Please include a summary of the changes and the related issue/task. Please also include relevant motivation and context. List any dependencies that are required for this change, including links to other pull requests/branches in other repositories if applicable.

Summary of changes: 
1. Ran `npm create vite@latest` and copied over the tsconfig, packages, eslint settings from a [generated Vite project](https://vitejs.dev/guide/#scaffolding-your-first-vite-project) 
2. Tweaked the config values with the `.src/` directory copied over into the boilerplate repo until I could run the Turalio scenario start-to-finish with all other repos on origin/dev (after version 1.0). 
3. Copied over some config values from request-generator's `vite.config.ts` and `package.json`. The repo's port is defined in the latter, so I removed the `PORT` env var in `.env`.
4. Removed in this repo's `package.json`'s any `scripts` referring to react-app-rewired and jest (vitest seems to use jest or a jest replacement under the hood, so all the method names are the same). 
5. Removed the `public/index.html` and replaced it with a root-level `index.html`; Vite doesn't use an `index.html` inside `public/`. https://vitejs.dev/guide/#index-html-and-project-root
6. Updated `prettier` and `typescript` packages
7. Moved TypeScript types, linting, testing packages over to `devDependencies`
8. Fixed [filename casing](https://stackoverflow.com/a/55541435) in git _locally_, so GitHub recognizes that cdsHooksCard.tsx and cdsHooksCards.tsx become CdsHooksCard.tsx and CdsHooksCards.tsx
9. Updated the GitHub Action from Node 12 to Node 21 so there will be helpful linting messages
10. Attempted to resolve a bunch of linting issues that popped up
11. Attempted to add types to the version of LForms we used, based off the current source code of [LForms](https://github.com/LHNCBC/lforms).

## Issue ticket number and Jira link

> Please include the Jira Ticket Number and Link for this issue/task.

N/A, this was done part of an effort to add more unit tests and mock network request responses (not done in this PR)

## Checklist before requesting a review

## Checklist for conducting a review

## Workflow 

Owner of the Pull Request will be responsible for merge after all requirements are met, including approval from at least one reviewer. Additional changes made after a review will dismiss any approvals and require re-review of the additional updates. Auto merging can be enabled below if additional changes are likely not to be needed. The bot will auto assign reviewers to your Pull Request for you. 

# How to manually test
1. Check out all other repositories on `dev`
2. Select Jon Snow & Turalio in request-generator
3. Click the "Send Rx to Pharmacy" button

## Check that the port is configurable in the `.env`
4. In the rems-smart-on-fhir's `.env`, change the port from 4040 to something else between 1-65535. Do `npm start`.
5. Click the "Launch SMART on FHIR app" button. You will notice the url will not resolve.
6. Do Ctrl+C on rems-smart-on-fhir and revert back to port 4040 for the next step.

## Rest of the Turalio scenario
7. Click the "Launch SMART on FHIR app" button.
8. Open the Network tab.
9. Send the patient-view hook (refresh button).
10. Send the order-select hook by selecting Turalio (or any other MedicationRequest) from the dropdown or clicking the refresh button
11. Fill out the Turalio Patient Enrollment, Turalio Prescriber Enrollment, and Turalio Knowledge Assessment forms. 
12. Resend the patient-view or order-select hook to check that the card for the Turalio Patient Status Update form comes back.
13. Click the 'Add "Completion of Patient Status Update Questionnaire" to Task List' button. If you do a GET request from the test-ehr on `http://localhost:8080/test-ehr/r4/Task`, you should see another Task.
14. Click the "Patient Status Update Form" button and open the Network tab. When the form loads you will see requests to different FHIR resources (including `$questionnaire-package`).
15. Modify the current form anywhere and click the "Save to EHR" button. Repeat again several times.
16. Wait 1 minute and click the "Load Previous Form" button or close the tab form in the top navigation bar and click the "Patient Status Update Form" button again. You should see the saved forms modal pop up. 
17. Select a saved form and click "Submit".
18. Check the ETASU status to confirm that the Patient Status Update Form was recorded.
